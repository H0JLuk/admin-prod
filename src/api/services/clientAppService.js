import { Api } from '../apiClient';
import { getReqOptions } from './index';

export async function getClientAppList() {
    return Api.get('/admin/clientApplication', getReqOptions());
}

export async function addClientApp(clientAppDto) {
    return Api.post('/admin/clientApplication', clientAppDto, getReqOptions());
}

export async function copyClientApp(clientAppDto) {
    return Api.post('/admin/clientApplication/copy', clientAppDto, getReqOptions());
}

export async function updateClientApp(id, clientAppDto) {
    return Api.put(`/admin/clientApplication/${id}`, clientAppDto, getReqOptions());
}

export async function deleteClientApp(id) {
    return Api.delete(`/admin/clientApplication/${id}`, getReqOptions());
}

export async function updateProperties(id, properties) {
    return Api.put(`/admin/clientApplication/${id}/properties`, properties, getReqOptions());
}

export async function reorderClientApp(idMap) {
    return Api.put('/admin/clientApplication/reorder', idMap, getReqOptions());
}

/** TODO: remove DEFAULT_APP and functions after second phase */
const DEFAULT_APP = 'DEFAULT_APP';

export function setDefaultAppCode(appCode) {
    sessionStorage.setItem(DEFAULT_APP, appCode);
}

export function getDefaultAppCode() {
    return sessionStorage.getItem(DEFAULT_APP) || '';
}

export function removeDefaultAppCode() {
    sessionStorage.removeItem(DEFAULT_APP);
}
