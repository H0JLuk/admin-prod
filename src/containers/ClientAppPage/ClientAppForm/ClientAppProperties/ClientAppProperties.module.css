.clientAppForm {
    position: relative;
}

.buttonGroup {
    margin-left: 15px;
}

.form {
    padding-bottom: 30px;
}

.container {
    background-color: #f8f8f8;
    padding: 20px;
    margin: 0 30px 15px 15px;
}

.container :global(.ant-row:last-child .ant-form-item) {
    margin-bottom: 0;
}

.row {
    margin-top: 10px;
    max-width: 950px;
}
