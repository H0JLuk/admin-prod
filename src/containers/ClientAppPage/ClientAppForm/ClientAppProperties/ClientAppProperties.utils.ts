import { APP_MECHANIC } from '@constants/clientAppsConstants';

export function computeDisabledMechanics(mechanics: APP_MECHANIC[]): APP_MECHANIC[] {
    if (!mechanics.length) {
        return [];
    }
    const isExpressExist = mechanics.includes(APP_MECHANIC.EXPRESS);
    const isMixExist = mechanics.includes(APP_MECHANIC.MIX);
    const isEcosystemExist = mechanics.includes(APP_MECHANIC.ECOSYSTEM);
    const isBundleExist = mechanics.includes(APP_MECHANIC.BUNDLE);

    const disabledMechanics: Record<keyof typeof APP_MECHANIC, boolean> = {
        [APP_MECHANIC.ECOSYSTEM]: isExpressExist || isMixExist,
        [APP_MECHANIC.PRESENTS]: isExpressExist,
        [APP_MECHANIC.PRESENTATION]: isExpressExist,
        [APP_MECHANIC.BUNDLE]: isExpressExist || isMixExist,
        [APP_MECHANIC.EXPRESS]: !isExpressExist,
        [APP_MECHANIC.MIX]: isExpressExist || isEcosystemExist || isBundleExist,
    };

    return (Object.keys(disabledMechanics) as APP_MECHANIC[]).reduce<APP_MECHANIC[]>((acc, key) => {
        if (disabledMechanics[key]) {
            acc.push(key);
        }
        return acc;
    }, []) as APP_MECHANIC[];
}
