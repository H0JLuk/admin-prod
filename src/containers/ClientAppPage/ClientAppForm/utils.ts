import { customNotifications } from '@utils/notifications';
import { ArgsProps } from 'antd/lib/notification';
import { addSettings, updateSettingsList } from '@apiServices/settingsService';
import { SETTINGS_TYPES } from './ClientAppFormConstants';
import { SettingDto } from '@types';
import ROLES from '@constants/roles';

export interface IChangedParam extends Pick<SettingDto, 'clientAppCode' | 'key' | 'value' | 'userRole'> {
    type: SETTINGS_TYPES;
}

export const showNotify = (message: React.ReactNode, isError?: boolean) => {
    const config: ArgsProps = {
        duration: 5,
        message,
    };

    if (isError) {
        customNotifications.error(config);
    } else {
        customNotifications.success(config);
    }

};

type UpdateAndAddSettings = {
    updateSettings: SettingDto[];
    addSettingsArr: SettingDto[];
};

export async function createOrUpdateKey(changedParams: IChangedParam[]) {
    const { updateSettings, addSettingsArr } = changedParams.reduce<UpdateAndAddSettings>((result, { type, ...params }) => {
        const key = type === SETTINGS_TYPES.EDIT ? 'updateSettings' : 'addSettingsArr';

        if (params.key === 'referralTokenLifetime') {
            delete params.clientAppCode;
            params.key = 'token_lifetime';
            params.userRole = ROLES.REFERAL_LINK;
        }

        return {
            ...result,
            [key]: [...result[key], params],
        };
    }, { updateSettings: [], addSettingsArr: [] });

    updateSettings.length && (await updateSettingsList(updateSettings));
    addSettingsArr.length && (await addSettings(addSettingsArr));
}
