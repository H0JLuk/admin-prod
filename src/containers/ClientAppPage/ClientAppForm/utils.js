import { notification } from 'antd';
import { addSettings, updateSettingsList } from '../../../api/services/settingsService';
import { designKeysForCheck, SETTINGS_TYPES } from './ClientAppFormConstants';

export const showNotify = (message, isError) => {
    const config = {
        duration: 5,
        placement: 'bottomRight',
    };

    if (isError) {
        notification.error({
            ...config,
            message,
        });
    } else {
        notification.success({
            ...config,
            message,
        });
    }

};

export async function createOrUpdateKey (changedParams) {
    const updateSettings = [];
    const addSettingsArr = [];

    changedParams.forEach(({ type, ...params }) => {
        type === SETTINGS_TYPES.EDIT
            ? updateSettings.push(params)
            : addSettingsArr.push(params);
    });

    updateSettings.length && (await updateSettingsList(updateSettings));
    addSettingsArr.length && (await addSettings(addSettingsArr));
}

export const checkExistDesignSettings = (settings) => {
    return designKeysForCheck.some(key => !Object.prototype.hasOwnProperty.call(settings, key));
};

export const getImage = (imageName) => {
    return `${process.env.PUBLIC_URL}/images/clientAppDesign/${imageName}.png`;
};


