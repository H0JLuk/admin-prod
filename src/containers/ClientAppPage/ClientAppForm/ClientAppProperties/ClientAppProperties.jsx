import React, { useEffect, useState } from 'react';
import { Button, Form } from 'antd';
import { useHistory } from 'react-router-dom';
import { createOrUpdateKey, showNotify } from '../utils';
import { addClientApp, updateClientApp } from '../../../../api/services/clientAppService';
import { getAppCode, saveAppCode } from '../../../../api/services/sessionService';
import { addSettings } from '../../../../api/services/settingsService';
import { CLIENT_APPS_PAGES } from '../../../../constants/route';
import {
    ADD_BUTTON_TITLE,
    BACKEND_ERROR_ALREADY_EXIST_ENDING,
    EDIT_BUTTON_LABEL,
    formElements,
    FORM_MODES,
    keysToString,
    mainInfoElements,
    SETTINGS_TYPES,
    SUCCESS_PROPERTIES_CREATE_DESCRIPTION,
} from '../ClientAppFormConstants';
import AppFormConstructor from '../FormConstructor/FormConstructor';
import Loading from '../../../../components/Loading/Loading';
import PrivacyPolicy from './PrivacyPolicy/PrivacyPolicy';

import styles from './ClientAppProperties.module.css';

const ClientAppProperties = ({
    type,
    matchPath,
    propertiesSettings: { current: propertiesSettings },
    consent,
    updateSettings,
}) => {
    const history = useHistory();
    const [form] = Form.useForm();
    const isEdit = type === FORM_MODES.EDIT;
    const [loading, setLoading] = useState(false);
    const [btnStatus, setBtnStatus] = useState(true);

    useEffect(() => {
        isEdit && form.setFieldsValue(propertiesSettings);
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    const handleSubmit = async (formData) => {
        try {
            if (!isEdit) {
                const {
                    code,
                    name,
                    mechanics,
                    displayName,
                    login_types,
                    notification_types,
                    ...restData
                } = formData;

                setLoading(true);
                await addClientApp({ code, displayName, isDeleted: false, name });

                const settings = [
                    {
                        clientAppCode: code,
                        value: JSON.stringify(mechanics || []),
                        key: 'mechanics',
                    },
                    {
                        clientAppCode: code,
                        value: JSON.stringify(login_types || []),
                        key: 'login_types',
                    },
                    {
                        clientAppCode: code,
                        value: JSON.stringify(notification_types || []),
                        key: 'notification_types',
                    },
                ];
                Object.keys(restData).forEach((key) => {
                    const value = restData[key];
                    value && settings.push({ clientAppCode: code, value, key });
                });
                await addSettings(settings);
                showNotify(SUCCESS_PROPERTIES_CREATE_DESCRIPTION);
                saveAppCode(code);
                history.replace(`${ matchPath }${CLIENT_APPS_PAGES.EDIT_APP}`);
            } else {
                const clientAppCode = getAppCode();
                const { displayName, code, name, ...restData } = formData;
                const changedParams = Object.keys(restData).reduce((result, key) => {
                    const valueFromServer = keysToString.includes(key)
                        ? JSON.stringify(propertiesSettings[key])
                        : propertiesSettings[key];
                    const valueInForm = keysToString.includes(key)
                        ? JSON.stringify(restData[key])
                        : restData[key];

                    if (valueFromServer === undefined && valueInForm) {
                        return [...result, { clientAppCode, key, value: valueInForm, type: SETTINGS_TYPES.CREATE }];
                    }

                    if ((!valueFromServer && valueInForm) || (valueFromServer && valueInForm !== valueFromServer)) {
                        return [...result, { clientAppCode, key, value: valueInForm, type: SETTINGS_TYPES.EDIT }];
                    }

                    return result;
                }, []);

                const { id } = propertiesSettings;
                const requests = [];
                const notifies = [];

                if (displayName !== propertiesSettings.displayName) {
                    requests.push(updateClientApp(id, { displayName, code, isDeleted: false, name }));
                    notifies.push(() => showNotify(
                        <span>
                            Отображаемое имя витрины изменено с <b>&quot;{ propertiesSettings.displayName }&quot;</b> на <b>&quot;{ displayName }&quot;</b>
                        </span>
                    ));
                }

                if (changedParams.length) {
                    requests.push(createOrUpdateKey(changedParams));
                    notifies.unshift(() => showNotify(
                        <span>
                            Настройки для витрины <b>&quot;{ displayName }&quot;</b> обновлены
                        </span>
                    ));
                }

                if (requests.length) {
                    setLoading(true);
                    await Promise.all(requests);

                    notifies.forEach(fn => fn());
                    setBtnStatus(true);
                    updateSettings({ ...formData, id });
                }
            }

        } catch ({ message }) {
            const messageToShow = message.endsWith(BACKEND_ERROR_ALREADY_EXIST_ENDING)
                ? `Клиентское приложение с кодом ${formData.code} уже существует `
                : message;
            showNotify(messageToShow, true);
        }
        setLoading(false);
    };

    const enableBtn = () => {
        setBtnStatus(false);
    };

    return (
        <div className={ styles.clientAppForm }>
            { loading && <Loading /> }
            <Form
                className={ styles.form }
                form={ form }
                onFinish={ handleSubmit }
                layout="vertical"
                onFieldsChange={ enableBtn }
                id={ type }
            >
                <div className={ styles.container }>
                    { mainInfoElements.map((row, index) => (
                        <AppFormConstructor
                            key={ index }
                            row={ row }
                            isEdit={ isEdit }
                        />
                    )) }
                </div>
                <div className={ styles.container }>
                    { formElements.map((row, index) => (
                        <AppFormConstructor key={ index } row={ row } />
                    )) }
                </div>
                <div className={ styles.container }>
                    <PrivacyPolicy consent={ consent } />
                </div>
                <div className={ styles.buttonGroup }>
                    <Button
                        disabled={ btnStatus }
                        type="primary"
                        htmlType="submit"
                    >
                        { !isEdit ? ADD_BUTTON_TITLE : EDIT_BUTTON_LABEL }
                    </Button>
                </div>
            </Form>
        </div>
    );
};

export default ClientAppProperties;
