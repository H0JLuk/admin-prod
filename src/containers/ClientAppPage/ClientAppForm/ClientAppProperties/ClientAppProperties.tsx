import React, { useEffect, useState } from 'react';
import { Button, Col, Form, FormProps, Row } from 'antd';
import { useHistory } from 'react-router-dom';
import { createOrUpdateKey, IChangedParam, showNotify } from '../utils';
import { addClientApp, updateClientApp } from '@apiServices/clientAppService';
import { getAppCode, saveAppCode } from '@apiServices/sessionService';
import { addSettings } from '@apiServices/settingsService';
import { CLIENT_APPS_PAGES } from '@constants/route';
import Loading from '@components/Loading';
import AppFormConstructor from '../FormConstructor';
import SelectTags from '@components/SelectTags';
import PrivacyPolicy from './PrivacyPolicy';
import {
    ADD_BUTTON_TITLE,
    BACKEND_ERROR_ALREADY_EXIST_ENDING,
    BUSINESS_ROLE_FOR_APPLICATION,
    BUSINESS_ROLE_FOR_APP_PLACEHOLDER,
    EDIT_BUTTON_LABEL,
    formElements,
    FORM_MODES,
    keysToString,
    mainInfoElements,
    SETTINGS_TYPES,
    SUCCESS_PROPERTIES_CREATE_DESCRIPTION,
} from '../ClientAppFormConstants';
import { LoginTypes, LOGIN_TYPES_ENUM } from '@constants/loginTypes';
import { NOTIFICATION_TYPES } from '@constants/clientAppsConstants';
import { IPropertiesSettings, ISettings } from '../ClientAppContainer';
import { BusinessRoleDto, ConsentDto, SettingDto } from '@types';
import { compareArrayOfNumbers } from '@utils/helper';

import styles from './ClientAppProperties.module.css';

type ClientAppPropertiesProps = {
    businessRoles: React.MutableRefObject<BusinessRoleDto[]>;
    type: FORM_MODES;
    matchPath: string;
    propertiesSettings: React.MutableRefObject<Partial<ISettings>>;
    updateSettings: (formData: IPropertiesSettings) => void;
    consent: ConsentDto | null;
};

const ClientAppProperties: React.FC<ClientAppPropertiesProps> = ({
    businessRoles: { current: businessRoles },
    type,
    matchPath,
    propertiesSettings: { current: propertiesSettings },
    consent,
    updateSettings,
}) => {
    const history = useHistory();
    const [form] = Form.useForm();
    const isEdit = type === FORM_MODES.EDIT;
    const [loading, setLoading] = useState(false);
    const [btnStatus, setBtnStatus] = useState(true);
    const [disabledFields, setDisabledFields] = useState({} as Record<string, string[]>);

    useEffect(() => {
        isEdit && form.setFieldsValue(propertiesSettings);
        updateCheckboxStatus(propertiesSettings.login_types as any);
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    const handleSubmit = async (formData: Record<string, any>) => {
        try {
            if (!isEdit) {
                const {
                    code,
                    name,
                    mechanics,
                    game_mechanics,
                    displayName,
                    login_types,
                    notification_types,
                    businessRoleIds,
                    ...restData
                } = formData;

                setLoading(true);
                await addClientApp({ code, displayName, isDeleted: false, name, businessRoleIds });

                const settings: SettingDto[] = [
                    {
                        clientAppCode: code,
                        value: JSON.stringify(mechanics || []),
                        key: 'mechanics',
                    },
                    {
                        clientAppCode: code,
                        value: JSON.stringify(game_mechanics || []),
                        key: 'game_mechanics',
                    },
                    {
                        clientAppCode: code,
                        value: JSON.stringify(login_types || []),
                        key: 'login_types',
                    },
                    {
                        clientAppCode: code,
                        value: JSON.stringify(notification_types || []),
                        key: 'notification_types',
                    },
                ];
                Object.keys(restData).forEach((key) => {
                    const value = restData[key];
                    value && settings.push({ clientAppCode: code, value, key });
                });

                await addSettings(settings);
                showNotify(SUCCESS_PROPERTIES_CREATE_DESCRIPTION);
                saveAppCode(code);
                history.replace(`${ matchPath }${CLIENT_APPS_PAGES.EDIT_APP}`);
            } else {
                const clientAppCode = getAppCode() || '';
                const {
                    displayName,
                    code,
                    name,
                    businessRoleIds,
                    ...restData
                } = formData;
                const changedParams = Object.keys(restData).reduce<IChangedParam[]>((result, key) => {
                    const valueFromServer = keysToString.includes(key)
                        ? JSON.stringify(propertiesSettings[key])
                        : propertiesSettings[key];
                    const valueInForm = keysToString.includes(key)
                        ? JSON.stringify(restData[key])
                        : restData[key];
                    const commonObj = { clientAppCode, key, value: valueInForm };

                    if (valueFromServer === undefined && valueInForm) {
                        return [...result, { ...commonObj, type: SETTINGS_TYPES.CREATE }];
                    }

                    if ((!valueFromServer && valueInForm) || (valueFromServer && valueInForm !== valueFromServer)) {
                        return [...result, { ...commonObj, type: SETTINGS_TYPES.EDIT }];
                    }
                    return result;
                }, []);

                const { id } = propertiesSettings;
                const requests: Promise<any>[] = [];
                const notifies: (() => void)[] = [];

                if (
                    displayName !== propertiesSettings.displayName ||
                    !compareArrayOfNumbers(businessRoleIds, propertiesSettings.businessRoleIds)
                ) {
                    requests.push(updateClientApp(id as any, { displayName, code, isDeleted: false, name, businessRoleIds }));
                    if (displayName !== propertiesSettings.displayName) {
                        notifies.push(() => showNotify(
                            <span>
                                Отображаемое имя витрины изменено с <b>&quot;{propertiesSettings.displayName}&quot;</b> на <b>&quot;{displayName}&quot;</b>
                            </span>,
                        ));
                    }

                    if (!compareArrayOfNumbers(businessRoleIds, propertiesSettings.businessRoleIds)) {
                        notifies.push(() => showNotify(
                            <span>
                                Бизнес-роли для витрины <b>&quot;{propertiesSettings.displayName}&quot;</b> успешно обновлены
                            </span>,
                        ));
                    }
                }

                if (changedParams.length) {
                    requests.push(createOrUpdateKey(changedParams));
                    notifies.unshift(() => showNotify(
                        <span>
                            Настройки для витрины <b>&quot;{displayName}&quot;</b> обновлены
                        </span>,
                    ));
                }

                if (requests.length) {
                    setLoading(true);
                    await Promise.all(requests);

                    notifies.forEach(fn => fn());
                    setBtnStatus(true);
                    updateSettings({ ...formData, id: id! });
                    setLoading(false);
                }
            }
        } catch ({ message }) {
            const messageToShow = message.endsWith(BACKEND_ERROR_ALREADY_EXIST_ENDING) ? (
                <span>
                    Клиентское приложение с кодом <b>{formData.code}</b> уже существует
                </span>
            ) : message;
            showNotify(messageToShow, true);
            setLoading(false);
        }
    };

    const updateCheckboxStatus = (value: LoginTypes[] = []) => {
        if (!value.length) {
            setDisabledFields({});
            return;
        }

        if (value.includes(LOGIN_TYPES_ENUM.DIRECT_LINK)) {
            setDisabledFields({
                notification_types: [NOTIFICATION_TYPES.PUSH],
                login_types: [LOGIN_TYPES_ENUM.PASSWORD, LOGIN_TYPES_ENUM.SBER_REGISTRY],
            });
            form.setFieldsValue({
                notification_types: [],
                login_types: [LOGIN_TYPES_ENUM.DIRECT_LINK],
            });
            return;
        }

        setDisabledFields({
            login_types: [LOGIN_TYPES_ENUM.DIRECT_LINK],
        });
    };

    const handleFieldsChange: FormProps['onFieldsChange'] = (fields) => {
        const [ { name, value } ] = fields;
        if (Array.isArray(name) && name[0] === 'login_types') {
            updateCheckboxStatus(value);
        }
        setBtnStatus(false);
    };

    return (
        <div className={styles.clientAppForm}>
            {loading && <Loading />}
            <Form
                className={styles.form}
                form={form}
                onFinish={handleSubmit}
                layout="vertical"
                onFieldsChange={handleFieldsChange}
                id={type}
            >
                <div className={styles.container}>
                    {mainInfoElements.map((row, index) => (
                        <AppFormConstructor
                            key={index}
                            row={row}
                            isEdit={isEdit}
                            isCreate={!isEdit}
                        />
                    ))}
                </div>
                <div className={styles.container}>
                    {formElements.map((row, index) => (
                        <AppFormConstructor
                            key={index}
                            row={row}
                            disabledFields={disabledFields}
                            isCreate={!isEdit}
                        />
                    ))}
                    <Row gutter={24} className={styles.row}>
                        <Col span={12}>
                            <Form.Item
                                label={BUSINESS_ROLE_FOR_APPLICATION}
                                name="businessRoleIds"
                                normalize={(catArr) => catArr.map(Number)}
                            >
                                <SelectTags
                                    data={businessRoles}
                                    idKey="id"
                                    placeholder={BUSINESS_ROLE_FOR_APP_PLACEHOLDER}
                                />
                            </Form.Item>
                        </Col>
                    </Row>
                </div>
                <div className={styles.container}>
                    <PrivacyPolicy consent={consent} />
                </div>
                <div className={styles.buttonGroup}>
                    <Button
                        disabled={btnStatus}
                        type="primary"
                        htmlType="submit"
                    >
                        {!isEdit ? ADD_BUTTON_TITLE : EDIT_BUTTON_LABEL}
                    </Button>
                </div>
            </Form>
        </div>
    );
};

export default ClientAppProperties;
