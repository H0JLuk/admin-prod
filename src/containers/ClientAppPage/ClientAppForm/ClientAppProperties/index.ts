import ClientAppProperties from './ClientAppProperties';

export default ClientAppProperties;
