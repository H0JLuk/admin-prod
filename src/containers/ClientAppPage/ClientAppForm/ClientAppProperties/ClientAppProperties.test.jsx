import React from 'react';
import { Form } from 'antd';
import { mount, shallow } from 'enzyme';
import { MemoryRouter } from 'react-router';
import {
    BACKEND_ERROR_ALREADY_EXIST_ENDING,
    FORM_MODES,
} from '../ClientAppFormConstants';
import { sleep } from '../../../../setupTests';
import { act } from 'react-dom/test-utils';
import ClientAppProperties from './ClientAppProperties';

import * as clientAppService from '../../../../api/services/clientAppService';
import * as settingsService from '../../../../api/services/settingsService';
import * as sessionService from '../../../../api/services/sessionService';
import * as clientAppPropertiesUtils from '../utils';

clientAppPropertiesUtils.showNotify = jest.fn();
clientAppService.addClientApp = jest.fn();
settingsService.addSettings = jest.fn();


describe('<ClientAppProperties /> test', () => {
    const props = {
        type: FORM_MODES.EDIT,
        matchPath: 'matchPath',
        propertiesSettings: {
            current: {
                token_lifetime: '10',
                inactivity_time: '10',
                tmp_block_time: '10',
                max_password_attempts: '10',
                max_presents_number: '10',
                name: 'name',
                code: 'code',
                installation_url: '124',
                login_types: ['PASSWORD', 'SBER_REGISTRY'],
                mechanics: ['PRESENTS'],
                privacyPolicy: 'privacyPolicy',
                ym_token: '558648281231',
                usage_url: '124',
                displayName: 'displayName',
                isDeleted: false,
            },
        },
        updateSettings: jest.fn(),
    };

    it('should match snapshot ', () => {
        const ClientAppPropertiesComponent = shallow(<ClientAppProperties { ...props } />);

        expect(ClientAppPropertiesComponent.debug()).toMatchSnapshot();
    });

    it('should properly create clientApp', async () => {
        await act(async () => {
            const newProps = { ...props, type: FORM_MODES.NEW };
            const ClientAppPropertiesComponent = mount(
                <MemoryRouter>
                    <ClientAppProperties { ...newProps } />
                </MemoryRouter>
            );

            ClientAppPropertiesComponent.find(Form)
                .props()
                .form.setFieldsValue(props.propertiesSettings.current);
            ClientAppPropertiesComponent.find(Form)
                .props()
                .form.submit();
            await sleep();

            expect(clientAppService.addClientApp).toHaveBeenCalledTimes(1);
            expect(clientAppService.addClientApp).toHaveBeenCalledWith({
                code: props.propertiesSettings.current.code,
                displayName: props.propertiesSettings.current.displayName,
                isDeleted: props.propertiesSettings.current.isDeleted,
                name: props.propertiesSettings.current.name,
            });
        });
    });

    it('should properly edit clientApp', async () => {
        await act(async () => {
            const newProps = {
                ...props,
                type: FORM_MODES.EDIT,
                propertiesSettings: {
                    current: {
                        ...props.propertiesSettings.current,
                        privacyPolicy: undefined,
                    },
                },
            };
            const ClientAppPropertiesComponent = mount(
                <MemoryRouter>
                    <ClientAppProperties { ...newProps } />
                </MemoryRouter>
            );

            clientAppService.addClientApp = jest.fn();
            settingsService.addSettings = jest.fn();
            clientAppPropertiesUtils.createOrUpdateKey = jest.fn();
            sessionService.getAppCode = jest.fn(() => 'appCode');

            await sleep();
            ClientAppPropertiesComponent.find(Form)
                .props()
                .form.setFieldsValue({
                    privacyPolicy: 'test',
                    token_lifetime: '33',
                });
            ClientAppPropertiesComponent.find(Form)
                .props()
                .form.submit();
            await sleep();

            expect(sessionService.getAppCode).toBeCalledTimes(1);

            expect(props.updateSettings).toBeCalledTimes(1);
            expect(props.updateSettings.mock.calls[0][0]).toHaveProperty('code', newProps.propertiesSettings.current.code);
            expect(props.updateSettings.mock.calls[0][0]).toHaveProperty('displayName', newProps.propertiesSettings.current.displayName);
            expect(props.updateSettings.mock.calls[0][0]).toHaveProperty('token_lifetime', '33');

            expect(clientAppPropertiesUtils.createOrUpdateKey).toBeCalledTimes(1);
        });
    });

    it('should properly handle errors', async () => {
        await act(async () => {
            const ClientAppPropertiesComponent = mount(
                <MemoryRouter>
                    <ClientAppProperties { ...props } />
                </MemoryRouter>
            );

            clientAppPropertiesUtils.createOrUpdateKey.mockRejectedValue({
                message: BACKEND_ERROR_ALREADY_EXIST_ENDING,
            });

            await sleep();
            ClientAppPropertiesComponent.find(Form)
                .props()
                .form.setFieldsValue({
                    privacyPolicy: 'test',
                    token_lifetime: '33',
                });
            ClientAppPropertiesComponent.find(Form)
                .props()
                .form.submit();
            await sleep();

            expect(clientAppPropertiesUtils.showNotify).toBeCalledTimes(1);
            expect(clientAppPropertiesUtils.showNotify).toBeCalledWith(
                `Клиентское приложение с кодом ${props.propertiesSettings.current.code} уже существует `,
                true
            );
        });
    });
});
