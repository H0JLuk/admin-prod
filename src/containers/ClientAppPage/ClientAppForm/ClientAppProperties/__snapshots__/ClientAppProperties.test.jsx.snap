// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ClientAppProperties /> test should match snapshot  1`] = `
"<div className=\\"clientAppForm\\">
  <ForwardRef(InternalForm) className=\\"form\\" form={{...}} onFinish={[Function: handleSubmit]} layout=\\"vertical\\" onFieldsChange={[Function: handleFieldsChange]} id=\\"edit\\">
    <div className=\\"container\\">
      <AppFormConstructor row={{...}} isEdit={true} />
      <AppFormConstructor row={{...}} isEdit={true} />
    </div>
    <div className=\\"container\\">
      <AppFormConstructor row={{...}} disabledFields={{...}} />
      <AppFormConstructor row={{...}} disabledFields={{...}} />
      <AppFormConstructor row={{...}} disabledFields={{...}} />
      <AppFormConstructor row={{...}} disabledFields={{...}} />
      <Row gutter={24} className=\\"row\\">
        <Col span={12}>
          <FormItem label=\\"Бизнес-роль, для которой доступно приложение\\" name=\\"businessRoleIds\\" normalize={[Function: normalize]}>
            <SelectTags data={{...}} idKey=\\"id\\" placeholder=\\"Выберите роль/роли\\" />
          </FormItem>
        </Col>
      </Row>
    </div>
    <div className=\\"container\\">
      <PrivacyPolicy consent={[undefined]} />
    </div>
    <div className=\\"buttonGroup\\">
      <Button disabled={true} type=\\"primary\\" htmlType=\\"submit\\">
        Сохранить
      </Button>
    </div>
  </ForwardRef(InternalForm)>
</div>"
`;
